Remora.Commands/Builders/CommandParameterBuilder.cs

```diff
@@ -0,0 +1,479 @@
+    /// </summary>
+    /// <returns>The constructed <see cref="IParameterShape"/>.</returns>
+    public IParameterShape Build()
+    {
+        var rangeAttribute = _attributes.OfType<RangeAttribute>().SingleOrDefault();
+        var optionAttribute = _attributes.OfType<OptionAttribute>().SingleOrDefault();
+
+        if (optionAttribute is SwitchAttribute)
+        {
+            if (_defaultValue is not bool || _parameterType != typeof(bool))
+            {
+                throw new InvalidOperationException
+                (
+                 $"A switch parameter must be a {typeof(bool)}" +
+                 $" (The parameter \"{_name}\" was marked as ({_parameterType})."
+                );
+            }
+        }
+
+        return optionAttribute is null
+                   ? CreatePositionalParameterShape(rangeAttribute, this)
+                   : CreateNamedParameterShape(optionAttribute, rangeAttribute, this);
+    }
+
+    private static IParameterShape CreateNamedParameterShape
+    (
+        OptionAttribute optionAttribute,
+        RangeAttribute? rangeAttribute,
+        CommandParameterBuilder builder
+    )
+    {
+        var isCollection = builder._parameterType!.IsSupportedCollection();
+
+        IParameterShape newNamedParameter;
+        if (optionAttribute is SwitchAttribute)
+        {
+            newNamedParameter = CreateNamedSwitchParameterShape(optionAttribute, builder);
+        }
+        else if (!isCollection)
+        {
+            newNamedParameter = builder._isGreedy
+                ? CreateNamedSingleValueParameterShape(optionAttribute, builder)
+                : CreateGreedyNamedSingleValueParameterShape(optionAttribute, builder);
+        }
+        else
+        {
+            newNamedParameter = CreateNamedCollectionParameterShape(optionAttribute, rangeAttribute, builder);
+        }
+
+        return newNamedParameter;
+    }
+
+    private static IParameterShape CreateNamedCollectionParameterShape
+    (
+        OptionAttribute optionAttribute,
+        RangeAttribute? rangeAttribute,
+        CommandParameterBuilder builder
+    )
+    {
+        var description = builder._description ?? Constants.DefaultDescription;
+
+        IParameterShape newNamedParameter = new NamedCollectionParameterShape
+        (
+         optionAttribute.ShortName,
+         optionAttribute.LongName,
+         rangeAttribute?.GetMin(),
+         rangeAttribute?.GetMax(),
+         builder._name,
+         builder._parameterType!,
+         builder._isOptional,
+         builder._defaultValue,
+         builder._attributes,
+         builder._conditions,
+         builder._index,
+         description
+        );
+
+        return newNamedParameter;
+    }
+
+    private static IParameterShape CreateNamedSwitchParameterShape
+    (
+        OptionAttribute optionAttribute,
+        CommandParameterBuilder builder
+    )
+    {
+        if (!builder._isOptional)
```

Comment:
Private fields should be just that - private. If you need external access, the access rights should be loosened or the value should be accessed another way. Accessing a private field like this is fine when it's an exception to the rule for good reason - here, it's probably better to alter the design such that you don't need to poke around in a class's internals.